Sources/PluginLibrary/NamingUtils.swift

```diff
@@ -10,8 +10,183 @@
+  // names = names.union(swiftKeywordsReservedInParticularContexts)
+  names.insert("Type")
+  names.insert("Protocol")
+
+  names = names.union(swiftKeywordsUsedInDeclarations)
+  names = names.union(swiftKeywordsUsedInStatements)
+  names = names.union(swiftKeywordsUsedInExpressionsAndTypes)
+  names = names.union(swiftCommonTypes)
+  names = names.union(swiftSpecialVariables)
+  return names
+}()
+
+private let reservedFieldNames: Set<String> = {
+  () -> Set<String> in
+  var names: Set<String> = []
+
+  // Properties are instance names, so can't shadow static class
+  // properties such as `protoMessageName`.
+
+  // Properties can't shadow methods.  For example, we don't need to
+  // avoid `isEqualTo` as a field name.
+
+  // Basic Message properties that we don't want to shadow
+  names.insert("isInitialized")
+  names.insert("unknownFields")
+
+  // Standard Swift property names we don't want
+  // to conflict with:
+  names.insert("debugDescription")
+  names.insert("description")
+  names.insert("dynamicType")
+  names.insert("hashValue")
+
+  // We don't need to protect all of these keywords, just the ones
+  // that interfere with type expressions:
+  // names = names.union(swiftKeywordsReservedInParticularContexts)
+  names.insert("Type")
+  names.insert("Protocol")
+
+  names = names.union(swiftKeywordsUsedInDeclarations)
+  names = names.union(swiftKeywordsUsedInStatements)
+  names = names.union(swiftKeywordsUsedInExpressionsAndTypes)
+  names = names.union(swiftCommonTypes)
+  names = names.union(swiftSpecialVariables)
+  return names
+}()
+
+/*
+ * Many Swift reserved words can be used as enum cases if we put
+ * backticks around them:
+ */
+private let quotableEnumCases: Set<String> = {
+  () -> Set<String> in
+  var names: Set<String> = []
+
+  // We don't need to protect all of these keywords, just the ones
+  // that interfere with enum cases:
+  // names = names.union(swiftKeywordsReservedInParticularContexts)
+  names.insert("associativity")
+  names.insert("dynamicType")
+  names.insert("optional")
+  names.insert("required")
+
+  names = names.union(swiftKeywordsUsedInDeclarations)
+  names = names.union(swiftKeywordsUsedInStatements)
+  names = names.union(swiftKeywordsUsedInExpressionsAndTypes)
+  // Common type and variable names don't cause problems as enum
+  // cases, because enum case names only appear in special contexts:
+  // names = names.union(swiftCommonTypes)
+  // names = names.union(swiftSpecialVariables)
+  return names
+}()
+
+/*
+ * Some words cannot be used for enum cases, even if they
+ * are quoted with backticks:
+ */
+private let reservedEnumCases: Set<String> = [
+  // Don't conflict with standard Swift property names:
+  "debugDescription",
+  "description",
+  "dynamicType",
+  "hashValue",
+  "init",
+  "rawValue",
+  "self",
+]
+
+
+private func isAllUnderscore(_ s: String) -> Bool {
+  if s.isEmpty {
+    return false
+  }
+  for c in s.characters {
```

Comment:
Nit: if you're concerned about performance, iterate over `s.unicodeScalars`, since I'm not sure if my character-literal speedup PR made it into Swift 3.1 😄